library/src/main/java/org/apache/apex/malhar/lib/fs/s3output/S3BlockUpload.java

```diff
@@ -0,0 +1,490 @@
+  {
+    processedBlocks = new HashSet<>();
+    recordData = new ArrayList<>();
+    currentWindowRecoveryState = new HashMap<>();
+    windowDataManager.setup(context);
+    s3Client = createClient();
+  }
+
+  /**
+   * Create AmazonS3 client using AWS credentials
+   * @return AmazonS3
+   */
+  protected AmazonS3 createClient()
+  {
+    AmazonS3 client = new AmazonS3Client(new BasicAWSCredentials(accessKey, secretAccessKey));
+    if (endPoint != null) {
+      client.setEndpoint(endPoint);
+    }
+    return client;
+  }
+
+  @Override
+  public void beginWindow(long windowId)
+  {
+    currentWindowId = windowId;
+    if (windowId <= windowDataManager.getLargestCompletedWindow()) {
+      replay(windowId);
+    }
+  }
+
+  /**
+   * Replay the state.
+   * @param windowId replay window Id
+   */
+  protected void replay(long windowId)
+  {
+    try {
+      @SuppressWarnings("unchecked")
+      Map<Long, UploadBlockMetadata> recoveredData = (Map<Long, UploadBlockMetadata>)windowDataManager.retrieve(windowId);
+      if (recoveredData == null) {
+        return;
+      }
+      for (Map.Entry<Long, UploadBlockMetadata> ubm: recoveredData.entrySet()) {
+        output.emit(ubm.getValue());
+        blockInfo.remove(ubm.getKey());
+      }
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public void endWindow()
+  {
+    if (recordData.size() > 0) {
+      processWaitBlocks();
+    }
+
+    for (Long blockId : processedBlocks) {
+      blockInfo.remove(blockId);
+    }
+    processedBlocks.clear();
+    if (currentWindowId > windowDataManager.getLargestCompletedWindow()) {
+      try {
+        windowDataManager.save(currentWindowRecoveryState, currentWindowId);
+      } catch (IOException e) {
+        throw new RuntimeException("Unable to save recovery", e);
+      }
+    }
+    currentWindowRecoveryState.clear();
+  }
+
+  @Override
+  public void teardown()
+  {
+    windowDataManager.teardown();
+  }
+
+  /**
+   * Process the blocks which are in wait state.
+   */
+  private void processWaitBlocks()
+  {
+    Iterator<AbstractBlockReader.ReaderRecord<Slice>> ite = recordData.iterator();
+
+    while (ite.hasNext()) {
+      AbstractBlockReader.ReaderRecord<Slice> blockData = ite.next();
+      if (blockInfo.get(blockData.getBlockId()) != null) {
+        uploadBlockIntoS3(blockData);
+        ite.remove();
+      }
+    }
+  }
+
+  /**
+   * Upload the block into S3 bucket.
+   * @param tuple block data
+   */
+  protected void uploadBlockIntoS3(AbstractBlockReader.ReaderRecord<Slice> tuple)
+  {
+    if (currentWindowId <= windowDataManager.getLargestCompletedWindow()) {
+      return;
+    }
+    S3BlockMetaData metaData = blockInfo.get(tuple.getBlockId());
+    if (metaData == null) {
+      recordData.add(tuple);
+      return;
+    }
+    if (processedBlocks.contains(tuple.getBlockId())) {
+      return;
+    }
+    processedBlocks.add(tuple.getBlockId());
```

Comment:
Yes we can do. I added this because of there is an issue in FSInputModule that the same block id is emitting. PR for this fix is merged. I removed the "processedBlocks" as it is not needed.